### PR TITLE
FHIR resources audit in OFS

### PR DESCRIPTION
**Pull request by ajitsonawane1**
The goal of this project is to create a utility that will take minimal inputs from the user and provide a detailed easy interpretation to get information about how source data gets ingested, transformed and reconciled into final FHIR resources.

**Before this PR(Utility):**

**For FHIR and CDA batch pipelines:**

```
1. Query operational-fhir-store to fetch DocumentReference ID  of source data file from GCS bucket.
2. Use this fetched DocumentReference ID as a reference to query provenance resource to get the details about transformed resources in the intermediate-fhir-store
3. To get the information about which pipeline/agent transformed source data to intermediate-fhir-store, you will get Device ID in the Provenance resource information which can be used to query Device resource and dataflow pipeline details.
4. To get information about reconciled resources in the final-fhir-store, you need to query the Provenance resource using intermediate-fhir-store resource ID.
```

**For CSV batch pipelines:**

```
1. Query operational-fhir-store to fetch DocumentReference ID  of source CSV data file from GCS bucket.
2. Since for CSV data harmonization, data first gets copied to Bigquery tables and from there it gets transformed into FHIR resources, get the details of DocumentReference ID  of Bigquery table using fetched DocumentReference ID  of source CSV data file.
3. Performed the same steps as for FHIR and CDA batch pipelines using DocumentReference ID of the Bigquery table.
4. Use this fetched DocumentReference ID of BQ table as a reference to query provenance resource to get the details about transformed resources in the intermediate-fhir-store.
5. To get the information about which pipeline/agent transformed source data to intermediate-fhir-store, you will get Device ID in the Provenance resource information which can be used to query Device resource and dataflow pipeline details.
6. To get information about reconciled resources in the final-fhir-store, you need to query the Provenance resource using intermediate-fhir-store resource ID.

```
`After following all the steps the user would be able to track only one single record from ingestion to reconciliation. There is no way to query multiple records for data lineage at the same time.`

`Additionally, users must be familiar with the relationships between FHIR resources like Provenance, DocumentReference, and Device to get the required details.`

**After this PR(Utility):**

With this utility the user needs to configure very minimal basic information in the config file as shown below. 

1. FILE_FORMAT = ""        #Source file format e.g. "csv", "ndjson", "xml"
2. GCS_FILE_PATH = ""   # Source file GCS Path e.g. "gs://pgt-csv-input/ehr2/batch_02/csvs/patients.csv"
3. HDE_PREFIX = ""          # HDE Prefix e.g. "pgt" 
4. HDE_ENV = ""               # HDE env where data transformed e.g. "stage", "synth", "dev", "prod" 
5. FHIR_STORE_LOC = "" # HDE FHIR store location e.g. "us", "us-central1", "us-east4" 
6. OUTPUT_FILE_PATH = "" # Local file path for storing the output e.g. "output.json", "/home/user/output.json" 

`Also, users don't need any information about FHIR resources like provenance, DocumentReference, and Device.`

`This will generate the following JSON output, which gives detailed information about the ingested data.`

image

### CODE REVIEW

**Comment by ajitsonawane1** [comment]
R: @svetakvsundhar

File: operations/fhir_resources_audit/main.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        return recon_results
+
+
+    def get_transformed_fhir_fhir_resources(self):
+        """Fetches how source records/data transformed into IFS FHIR resources and by which pipeline"""
+
+        for index, source_file_info in enumerate(self.results):
+            ifs_resources = {}
+            provenance_url = Util.get_provenance_resource_path_url(self.FHIR_URL, source_file_info['ingested_tbl_info']['id'])
+            print(provenance_url)
```
Comment (svetakvsundhar) [inline operations/fhir_resources_audit/main.py:66]: used for debugging? remove this print statement?

Comment (svetakvsundhar) [inline operations/fhir_resources_audit/main.py:66]: used for debugging? remove this statement?

File: operations/fhir_resources_audit/main.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+        return recon_results
+
+
+    def get_transformed_fhir_fhir_resources(self):
+        """Fetches how source records/data transformed into IFS FHIR resources and by which pipeline"""
+
+        for index, source_file_info in enumerate(self.results):
+            ifs_resources = {}
+            provenance_url = Util.get_provenance_resource_path_url(self.FHIR_URL, source_file_info['id'])
+            print(provenance_url)
```
Comment (svetakvsundhar) [inline operations/fhir_resources_audit/main.py:144]: same as above

File: operations/fhir_resources_audit/main.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+                     temp_ifs_resource_info['reference'] = ifs_resource['reference']
+
+                     #fetch reconcilation info for this resource
+                     temp_ifs_resource_info['reconciled_into'] = self.get_reconcilated_resources(ifs_resource['reference'])
+                     ifs_resource_info['transormed_resources'].append(temp_ifs_resource_info)
+                ifs_resource_info['transformed_by'] = resources['entry'][i]['resource']['agent'][0]['who']
+                self.results[index]['resources_transformed_ifs'].append(ifs_resource_info)
+
+    def process(self):
+        print("processing OFS resources for C-CDA to FHIR harmonization")
```
Comment (svetakvsundhar) [inline operations/fhir_resources_audit/main.py:169]: nit: use `fstring` to stay consistent

File: operations/fhir_resources_audit/main.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        provenance_recon_url = Util.get_reconcilated_resources_url(self.FHIR_URL, resource_ref)
+        resources = self.AUTH_QUERY.executeQuery(next_page_url=provenance_recon_url)
+        recon_results={}
+        recon_results['reconciled_resource'] = resources['entry'][0]['resource']['target']
+        recon_results['reconciled_by'] = resources['entry'][0]['resource']['agent'][0]['who']
+        return recon_results
+
+
+    def get_transformed_fhir_fhir_resources(self):
+        """Fetches how source records/data transformed into IFS FHIR resources and by which pipeline"""
```
Comment (svetakvsundhar) [inline operations/fhir_resources_audit/main.py:215]:
```suggestion
        """Fetches the specific source records/data and respective pipeline of transformed into IFS FHIR resources."""
```

**Review by svetakvsundhar — CHANGES_REQUESTED** [review]
Thanks for the great work @ajitsonawane1! A general comment, the `sample_output.json` takes up 247k lines of code here. Is it possible that we upload it as a screenshot, or an abridged version into the repo? Totally open to other suggestions.

File: operations/fhir_resources_audit/main.py (hunk 1 above)
Comment (ajitsonawane1) [reply operations/fhir_resources_audit/main.py]: Done

File: operations/fhir_resources_audit/main.py (hunk 2 above)
Comment (ajitsonawane1) [reply operations/fhir_resources_audit/main.py]: Done

File: operations/fhir_resources_audit/main.py (hunk 3 above)
Comment (ajitsonawane1) [reply operations/fhir_resources_audit/main.py]: Done

**Review by rachelgagnon — APPROVED** [review]
great work, thanks for this @ajitsonawane1! this is a super helpful utility.

**Review by surjits254 — APPROVED** [review]
LGTM

**Comment by ajitsonawane1** [comment]
> Thanks for the great work @ajitsonawane1! A general comment, the `sample_output.json` takes up 247k lines of code here. Is it possible that we upload it as a screenshot, or an abridged version into the repo? Totally open to other suggestions.

@svetakvsundhar Done.

**Comment by svetakvsundhar** [comment]
Thanks @ajitsonawane1 , this LGTM and I will work on merging it. As a follow up PR, could you add some unit tests and integration tests? That will be a good best practice to ensure correctness of the utility.

**Pull request merged**